### PR TITLE
UCP/RMA/TAG: Support fast-path short with new protocols

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: src/ucp/proto/proto_select.h [hunk 1]
```diff
@@ -61,6 +61,19 @@ typedef struct {
 } ucp_proto_select_t;
 
 
+/*
+ * Settings for short protocol
+ */
+typedef struct {
+    ssize_t             max_length_host_mem; /* max length of short protocol for
+                                                host memory buffer */
+    ssize_t             max_length_unknown_mem; /* max length of short protocol
```
Comment (dmitrygx) [inline src/ucp/proto/proto_select.h:70]: unaligned

File: src/ucp/proto/proto_select.c [hunk 2]
```diff
@@ -892,3 +893,72 @@ void ucp_proto_select_param_str(const ucp_proto_select_param_t *select_param,
...
+                            ucp_operation_id_t op_id, uint32_t op_attr_mask,
+                            unsigned proto_flags,
+                            ucp_proto_select_short_t *proto_short)
+{
+    ucp_context_h context    = worker->context;
+    const ucp_proto_t *proto = NULL;
+    const ucp_proto_threshold_elem_t *thresh;
+    ucp_proto_select_param_t select_param;
+    const ucp_proto_single_priv_t *spriv;
+    unsigned op_attr;
```
Comment (dmitrygx) [inline src/ucp/proto/proto_select.c:918]: `uint32_t` to be the same as for `op_attr_mask`?

File: src/ucp/proto/proto_select.c [hunk 3]
```diff
@@ -892,3 +893,72 @@ void ucp_proto_select_param_str(const ucp_proto_select_param_t *select_param,
...
+            if ((proto != thresh->proto_config.proto) ||
+                (proto_short->lane != spriv->super.lane) ||
+                (proto_short->rkey_index != spriv->super.rkey_index)) {
+                /* not all op_attr options have same configuration */
+                goto out_disable;
+            }
+
+            /* Fast-path threshold is the minimal of all op_attr options */
+            if (thresh->max_msg_length < proto_short->max_length_host_mem) {
+                proto_short->max_length_host_mem = thresh->max_msg_length;
```
Comment (dmitrygx) [inline src/ucp/proto/proto_select.c:949]: `proto_short->max_length_host_mem = ucs_min(proto_short->max_length_host_mem, thresh->max_msg_length);`
wdyt?

File: src/ucp/proto/proto_select.c [hunk 4]
```diff
@@ -892,3 +893,72 @@ void ucp_proto_select_param_str(const ucp_proto_select_param_t *select_param,
...
+        }
+    }
+
+    /* If we support only host memory, set max short for unknown memory type to
+     * be same as for host memory type. Otherwise, disable short if memory type
+     * is unknown.
+     */
+    proto_short->max_length_unknown_mem = (context->num_mem_type_detect_mds > 0) ?
+                                          -1 : proto_short->max_length_host_mem;
+    return;
```
Comment (dmitrygx) [inline src/ucp/proto/proto_select.c:960]: what if `proto_short->max_length_host_mem == -1`? shat we disable short?
if we get here and `proto_short->max_length_host_mem > -1` has to be always true, let's add an assertion here to make sure

File: src/ucs/sys/math.h [hunk 5]
```diff
@@ -148,11 +148,33 @@ static inline double ucs_log2(double x)
...
+         _submask = 0; \
+         /* end when reaching _mask + 1 */ \
+         _submask <= (_mask); \
+         /* Increment _submask by 1. If it became larger than _mask, do nothing \
+          * here, and next condition check will exit the loop. Otherwise, add \
+          * ~mask to fast-forward the carry (from ++ operation) to the next \
+          * valid bit in _mask, and then do "& _mask" to remove any bits which \
+          * are not in the mask. \
+          */ \
+         _submask++, \
```
Comment (dmitrygx) [inline src/ucs/sys/math.h:174]: `(_submask)`
all occurrences

File: test/gtest/ucs/test_math.cc [hunk 6]
```diff
@@ -203,6 +203,34 @@ UCS_TEST_F(test_math, for_each_bit) {
...
+    /* Generate mask values to test */
+    std::vector<int64_t> masks;
+    masks.push_back(0);
+    masks.push_back(1);
+    masks.push_back(65536);
+    for (int i = 0; i < 100; ++i) {
+        masks.push_back(ucs::rand() % 65536);
+    }
+
+    for (std::vector<int64_t>::iterator iter = masks.begin();
```
Comment (dmitrygx) [inline test/gtest/ucs/test_math.cc:216]: `std::vector<int64_t>::const_iterator`, sicne the code doens't change a content referenced by `iter`

File: test/gtest/ucs/test_math.cc [hunk 7]
```diff
@@ -203,6 +203,34 @@ UCS_TEST_F(test_math, for_each_bit) {
...
 }
 
+UCS_TEST_F(test_math, for_each_submask) {
+    /* Generate mask values to test */
+    std::vector<int64_t> masks;
+    masks.push_back(0);
+    masks.push_back(1);
+    masks.push_back(65536);
+    for (int i = 0; i < 100; ++i) {
+        masks.push_back(ucs::rand() % 65536);
```
Comment (dmitrygx) [inline test/gtest/ucs/test_math.cc:213]: `masks.push_back(2 + (ucs::rand() % 65535))` to not test `0`/`1` several times?

File: src/ucp/rma/rma_send.c [hunk 8]
```diff
@@ -207,6 +207,31 @@ ucs_status_ptr_t ucp_put_nb(ucp_ep_h ep, const void *buffer, size_t length,
...
 
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_put_send_short(ucp_ep_h ep, const void *buffer, size_t length,
+                   uint64_t remote_addr, ucp_rkey_h rkey,
+                   const ucp_request_param_t *param)
+{
+    const ucp_rkey_config_t *rkey_config;
+    uct_rkey_t tl_rkey;
+
+    if (ucs_unlikely(param->op_attr_mask & (UCP_OP_ATTR_FIELD_DATATYPE |
```
Comment (brminich) [inline src/ucp/rma/rma_send.c:218]: Isn't it `INVALID_PARAM`?

File: src/ucp/proto/proto_select.c [hunk 9]
```diff
@@ -892,3 +893,77 @@ void ucp_proto_select_param_str(const ucp_proto_select_param_t *select_param,
...
+                            ucp_proto_select_short_t *proto_short)
+{
+    ucp_context_h context    = worker->context;
+    const ucp_proto_t *proto = NULL;
+    const ucp_proto_threshold_elem_t *thresh;
+    ucp_proto_select_param_t select_param;
+    const ucp_proto_single_priv_t *spriv;
+    uint32_t op_attr;
+
+    ucs_for_each_submask(op_attr, op_attr_mask) {
```
Comment (brminich) [inline src/ucp/proto/proto_select.c:920]: why is this needed? Can you please give an example of set of flags?

File: src/ucp/rma/rma_send.c (hunk 8 above)
Comment (yosefe) [reply src/ucp/rma/rma_send.c:218]: it's UCS_ERR_NO_RESOURCE in order to fallback to using UCP request

File: src/ucp/proto/proto_select.c (hunk 9 above)
Comment (yosefe) [reply src/ucp/proto/proto_select.c:920]: added comment

**Comment by yosefe** [comment]
@brminich fixed

**Review by brminich — APPROVED** [review]

File: src/ucp/rma/rma_send.c (hunk 8 above)
Comment (brminich) [reply src/ucp/rma/rma_send.c:218]: yes, but not `UCP_OP_ATTR_FIELD_DATATYPE`

Comment (yosefe) [reply src/ucp/rma/rma_send.c:218]: if UCP_OP_ATTR_FIELD_DATATYPE is set, then it may not be contig, so must check datatype explicitly (for example by allocating request flow)
we could do same as 'tag' here - check if datatype==generic as optimization - but seems not needed for now since most users of this API do not set datatype field at all

Comment (brminich) [reply src/ucp/rma/rma_send.c:218]: but it is not really supported now..

Comment (yosefe) [reply src/ucp/rma/rma_send.c:218]: > but it is not really supported now..

Right, but it's not directly related to the code above, here we just disable fast-path short if datatype is specified

Comment (brminich) [reply src/ucp/rma/rma_send.c:218]: ok
